Various ports: add support for Timer hard= option

arachsys opened the pull request:
### Summary

7816b1f, 8f85eda and 8162468 (PR #17363) added a `hard=` option to the rp2 `Timer` constructor and initialiser to allow the callback to be switched between hard-IRQ and soft-IRQ mode.

The behaviour of other ports varies: either hard-IRQ only or soft-IRQ only. @dpgeorge suggested that it would be good to make this consistently configurable across all ports, preserving the current default but allowing both behaviours everywhere.

This patch series is an attempt to do that, adding support to every port that has a `machine.Timer` except nrf (see below), ensuring there is cross-port test coverage of both variants, and documenting as appropriate.

### Notes and issues

nrf is the only port with a `machine.Timer` that isn't covered by this patch series. This port doesn't have `mp_sched_schedule()` enabled so I don't think I can support `hard=False` there? I've left it out altogether, assuming that extra code size isn't worth it just to add an ignored `hard=True` option for consistency. (I also note that nrf doesn't support the `freq=` option to the `Timer` constructor/init, just `period=`, so it's already a cut-back timer implementation compared to other ports.)

I've added hard IRQ support to `shared/runtime/softtimer`, but have gated it to avoid bloat on platforms which have their own timer infrastructure and don't need hard IRQ support from the software timer. alif, mimxrt and samd are the three ports which turn this on.

### Testing

Although I have compile tested every patch and every port via the Micropython test infrastructure, I don't have hardware to test ports other than rp2. The generalised test I have written in `tests/extmod/machine_timer.py` would exercise old and new codepaths on every modified platform if run there.

I would greatly appreciate help in verifying that this series does indeed work correctly on every port it claims to cover! Obviously I will happily fix any issues that do show up.

For the purposes of writing tests:
- every port with `machine.Timer` except nrf supports both `hard=True` and `hard=False`
- every port with `machine.Timer` except nrf supports `freq=N` (`N` in Hz)
- every port with `machine.Timer` except renesas-ra supports `period=N` (`N` in ms)
- esp32 and esp8266 don't support software timers (`id = -1`)

To try to cover the various cases a bit more thoroughly, my new test covers all ports with a `machine.Timer` including esp32/esp8266 but excluding nrf, exercising hard and soft timers with varying `freq=` and explicitly checking that the heap is locked in hard callbacks and unlocked in soft callbacks. (The preexisting `tests/extmod/machine_soft_timer.py` already covers all the functionality available on nrf, and tests period= instead of freq=.)

### Additional consolidation

As suggested by @andrewleech, I've added an additional (somewhat bolder!) patch to show what consolidating the hard/soft callback from mp_irq_handler() with the various timer implementations would look like. The test suite suggests this gives small binary size improvements on ports like stm32 and rp2 as well as having a pleasingly negative diffstat.

andrewleech commented on ports/esp32/machine_timer.c:
```diff
@@ -156,8 +157,28 @@ static void machine_timer_isr(void *self_in) {
 static void machine_timer_isr_handler(machine_timer_obj_t *self) {
-    mp_sched_schedule(self->callback, self);
-    mp_hal_wake_main_task_from_isr();
+    if (self->ishard) {
+        // When executing code within a handler we must lock the scheduler to
+        // prevent any scheduled callbacks from running, and lock the GC to
+        // prevent any memory allocations.
+        mp_sched_lock();
+        gc_lock();
+        nlr_buf_t nlr;
+        if (nlr_push(&nlr) == 0) {
+            mp_call_function_1(self->callback, MP_OBJ_FROM_PTR(self));
+            nlr_pop();
+        } else {
+            // Uncaught exception; disable the callback so it doesn't run again.
+            self->repeat = 0;
+            mp_printf(MICROPY_ERROR_PRINTER, "uncaught exception in timer callback\n");
+            mp_obj_print_exception(MICROPY_ERROR_PRINTER, MP_OBJ_FROM_PTR(nlr.ret_val));
+        }
+        gc_unlock();
+        mp_sched_unlock();
+    } else {
+        mp_sched_schedule(self->callback, MP_OBJ_FROM_PTR(self));
+        mp_hal_wake_main_task_from_isr();
+    }
```
Perhaps this common snippet could be consolidated into the extmod machine timer C file for reuse by all the ports.

Though arguably it could be used by any hardware peripheral interrupts so perhaps it should be somewhere even more common?

arachsys replied:
I originally hoped to consolidate all these stanzas into a single place, ideally merging into the similar IRQ object too, and I have provided a shared implementation of this in the patch to `extmod/machine_timer.c`.

However, many ports roll their own and don't use `extmod/machine_timer.c` at all, and there are a lot of incidental differences between these ports. For example, in the esp32 patch you quote above, we have to `mp_hal_wake_main_task_from_isr()` after `mp_sched_schedule()`, unlike any other port. Some ports (renesas-ra, stm32) have the notion of channels which change the shape of the error messages. Different ports have different structures for the timer object with different ways to disable callbacks on uncaught exceptions.

I expect it would be possible to pull out something sufficient generalised like

```c
    if (hard) {
        mp_sched_lock();
        gc_lock();
        nlr_buf_t nlr;
        if (nlr_push(&nlr) == 0) {
            mp_call_function_1(callback, MP_OBJ_FROM_PTR(argument));
            nlr_pop();
        } else {
            /* caller has to disable and print exception */
            return nlr.ret_val;
        }
        gc_unlock();
        mp_sched_unlock();
    } else {
        mp_sched_schedule(callback, MP_OBJ_FROM_PTR(argument));
    }
    return NULL;
```

and then use it as

```c
    ret_val = dispatch_callback(self->callback, self);
    if (ret_val) {
        self->repeat = 0;
        mp_printf(MICROPY_ERROR_PRINTER, "uncaught exception in timer callback\n");
        mp_obj_print_exception(MICROPY_ERROR_PRINTER, MP_OBJ_FROM_PTR(ret_val));
    }
```

or something like that. If that covered the IRQ case as well as the Timer case for each port, it might give us a slight code shrinkage too, albeit slightly slower at interrupt time.

(I think I'm assuming that callbacks can't raise a VM abort in the above, or at least ignoring the exception if they do, by assuming `nlr.ret_val` is non-NULL in an exception return. Possibly we're doing that already by calling mp_obj_print_exception with them?)

arachsys replied:
If we're being really brave about eliminating repetition here: where ports roll their own soft-timers instead of using the standard implementation (esp32, rp2, etc)... should they? Or, once hard=True and hard=False are properly supported on the shared `extmod/machine_timer.c` implementation, should everything 'custom' be replaced by that? Probably a bit controversial for this PR though! :)

andrewleech replied:
Yeah it's always hard to know how far to take a change like this! 

FWIW most ports already have a [stub `mp_hal_wake_main_task_from_isr` ](https://github.com/micropython/micropython/blob/e993f53877027bc46b226d026845e523c2cc94ca/ports/rp2/mphalport.h#L80 )for exactly this kind of consolidation.

arachsys replied:
Okay, you've emboldened me! I've pushed an additional patch as a suggestion for how we might consolidate this, factoring out mp_irq_dispatch() as a generalised version of mp_irq_handler() and then using it throughout the timer implementations. Hopefully this should give a small size reduction on quite a few ports.

andrewleech commented on tests/extmod/machine_timer.py:
```diff
@@ -0,0 +1,51 @@
+    # on the esp32 and esp8266 ports. Timer ID 0 is guaranteed to exist.
+    timer_id = 0
+else:
+    timer_id = -1
```
It would be good to test with both hardware and software timers if possible on platforms that support it?

arachsys replied:
Yes, it'd be great to make this more general to cover both on platforms that support both. Perhaps I can
```python
try:
  t = machine.Timer(0)
  # hardware timers available; test them
except:
  # hardware timers not available
```
and ditto for `Timer(-1)` for the soft timers. Is there anywhere where timer ID 0 is available but unsafe to use?

arachsys replied:
As you suggested, I've added an extra test to cover hardware timers on the platforms that support them. I think `timer_id = 0` is the only choice that's guaranteed to be available on esp32 and esp8266, whereas `timer_id = 1` is a safe choice on stm32.

andrewleech reviewed (commented):
Great work broadening the timer features and port consistency here!

andrewleech commented on ports/esp32/machine_timer.c:
```diff
@@ -238,6 +246,7 @@ static mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n
     self->repeat = args[ARG_mode].u_int;
     self->handler = machine_timer_isr_handler;
     self->callback = args[ARG_callback].u_obj;
+    self->ishard = args[ARG_hard].u_bool;
```
It might be just me but I read `ishard` as "I shard", personally I'd go for `is_hard` or just `hard`. Minor detail though...

andrewleech commented on shared/runtime/softtimer.c:
```diff
@@ -88,7 +89,11 @@ void soft_timer_handler(void) {
-            mp_sched_schedule(entry->py_callback, MP_OBJ_FROM_PTR(entry));
+            if (mp_irq_dispatch(entry->py_callback, MP_OBJ_FROM_PTR(entry),
+                entry->flags & SOFT_TIMER_FLAG_HARD_CALLBACK)) {
+                // Uncaught exception; disable the callback so it doesn't run again.
```
Looks like the callback disable is missing now?

arachsys replied:
It's the line below that comment where I set entry->mode to SOFT_TIMER_MODE_ONE_SHOT.

andrewleech commented on shared/runtime/softtimer.h:
```diff
@@ -30,6 +30,7 @@
 
 #define SOFT_TIMER_FLAG_PY_CALLBACK (1)
 #define SOFT_TIMER_FLAG_GC_ALLOCATED (2)
+#define SOFT_TIMER_FLAG_HARD_CALLBACK (4)
```
Was 4 used here for a reason? Is 3 used elsewhere?

arachsys replied:
These are bit masks that get ORed together, so powers of two.

andrewleech reviewed (commented):
The consolidation looks good to me!

arachsys commented:
> The consolidation looks good to me!

Yes, I like it too to be honest. The size reduction was better than I expected. Thanks for encouraging me not to be too timid!

arachsys commented:
[Rebased to HEAD of master.]

dpgeorge commented:
Thanks for the contribution!  Always good to clean up things and consolidate across ports.

Unfortunately esp32 cannot support hardware interrupts, because the hardware interrupt does not run on the main thread.  The interrupt handler needs to acquire the GIL which is non-trivial.  Testing the `extmod/machine_hard_timer.py` test on ESP32 it indeed crashes running the `hard=True` part of the test, and the REPL locks up.

So, I think for esp32 just raise an exception if `hard=True` is passed in.

For esp8266, the `tests/extmod/machine_hard_timer.py` doesn't currently work because esp8266 doesn't support passing in arguments to the constructor.  Instead it needs to do `timer=Timer(0)` then `timer.init(...)`.  Might be worth a separate PR to make that work the same as other ports?

I didn't check other boards yet.

dpgeorge commented:
I tested PYBV10 (stm32 board).  This port doesn't support non-soft `machine.Timer`, ie the `id` must be `-1` on this port, so `machine_hard_timer.py` test fails.

Also, `machine_timer.py` fails because the `modes` and `kinds` dicts evaluate to a different order than what they are defined in (eg it does periodic before one-shot).  Probably needs to use a tuple for those to ensure consistent order across all ports.

dpgeorge commented on tests/extmod/machine_timer.py:
```diff
@@ -0,0 +1,51 @@
+    print("SKIP")
+    raise SystemExit
+
+if sys.platform in ("nrf"):
```
This won't work, it should be `sys.plaftorm in ("nrf",)`.

dpgeorge commented:
> Are there two versions of the pyboard, one of which supports hard timers and one of which does not?

They all support `pyb.Timer` which is the stm32 specific way to do hardware timers.  For `machine.Timer`, only soft timer is supported by stm32 at the moment.

This is starting to dig up the main issue that the `machine.Timer` class is not well specified nor very consistent across ports, see #2971 for that discussion.

Ideally we'd address #2971 first, and then work on implementing it.  But --for the sake of making progress -- this PR is good.  It just might end up being a bit imperfect due to the incomplete nature of the `Timer` class on various ports.

> While I'm doing this, what do you think about the final 'consolidation' patch?

Yes, that's quite nice.  Always good to factor code, eliminate duplication, and reduce code size!  Eventually we might want to add `Timer.irq` (like eg `Pin.irq` and `UART.irq`) and then it would use the standard `mpirq.c` interface/code and not need any refactoring like that.  But -- again in the interest of making progress -- your refactoring is good and I think we should do it.  So feel free to expose `mp_irq_dispatch()` in an earlier commit and then use it immediately when updating each port.

arachsys committed:
shared/runtime/mpirq: Factor out mp_irq_dispatch() and use it.

Separate out a routine to call an arbitrary function with arbitrary
argument either directly as a hard-IRQ handler or scheduled as a soft-IRQ
handler, adjusting mp_irq_handler() to wrap this. This can then be used
to implement other hard/soft callbacks, such as for machine.Timer.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
shared/runtime/softtimer: Add support for hard callbacks.

Add a flag SOFT_TIMER_HARD_CALLBACK to request that a soft timer's python
callback is run directly from the IRQ handler with the scheduler and heap
locked, instead of being scheduled via mp_sched_schedule().

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
extmod/machine_timer: Support hard IRQ soft timer callbacks.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

As on the rp2 port, add support to the generic software timer for a hard=
argument to explicitly choose between these, setting the default to False
to match the existing behaviour. This enables hard timer callbacks for
the alif, mimxrt and samd ports.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
rp2/machine_timer: Use mp_irq_dispatch() to reduce duplication.

Now that mp_irq_dispatch() is available to dispatch arbitary hard/soft
callbacks, take advantage of this for rp2 machine.Timer. This should
slightly reduce binary size.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
renesas-ra/timer: Support soft IRQ timer callbacks.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

As on the rp2 port, add support to the renesas-ra port for a hard= argument
to explicitly choose between these, setting the default to True to match
the existing behaviour.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
stm32/timer: Support soft IRQ timer callbacks.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

As on the rp2 port, add support to the stm32 port for a hard= argument
to explicitly choose between these, setting the default to True to match
the existing behaviour.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
zephyr/machine_timer: Support hard IRQ timer callbacks.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

As on the rp2 port, add support to the zephyr port for a hard= argument
to explicitly choose between these, setting the default to False to match
the existing behaviour.

Adjust the stack-limit check to use the ISR stack while the callback is
dispatched so that hard IRQ callbacks work, as with machine_pin.c and
machine_i2c_target.c IRQ callbacks.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
esp8266/modmachine: Support hard IRQ timer callbacks.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

As on the rp2 port, add support to the esp8266 port for a hard= argument
to explicitly choose between these, setting the default to False to match
the existing behaviour. Open-code this as we don't link against mpirq.c
so can't use mp_irq_dispatch().

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
esp32/machine_timer: Warn that hard timers are not implemented.

machine.Timer() has inconsistent behaviour between ports: some run
callbacks in hard IRQ context whereas others schedule them like soft IRQs.

Most ports now support a hard= argument to the machine.Timer constructor
or initialiser to explicitly choose between these behaviours. However,
esp32 does not support hardware interrupts because they are not delivered
to the main thread, so the interrupt handler would need to acquire the GIL.

Raise a ValueError if hard=True is requested for esp32 machine.Timer.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
tests: Generalise rp2 timer test into a cross-port test.

Now all ports with machine.Timer except nrf support both hard and
soft callbacks, generalise tests/ports/rp2_machine_timer.py into
tests/extmod/machine_timer.py.

There is an existing machine_soft_timer.py which varies period= and
covers the nrf port but skips esp32/esp8266 because they don't support
software timers. In our new test, we try varying freq= instead of period=,
and cover esp32/esp8266 (with a fixed choice of hardware timer) but skip
nrf because it doesn't support hard= or freq=.

Add a check that the heap is locked (so allocation fails) in hard
callbacks and it is unlocked (so allocation succeeds) in soft callbacks,
to ensure we're getting the right kind of callback, not falling back to
the default.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
tests: Test hardware timers as well as software timers.

On platforms where hardware timers are available, test these in each
combination of hard/soft and one-shot/periodic in the same way as for
software timers. Where a platform supports both software (id = -1) and
hardware (id >= 0) timers, the behaviour of both is now checked.

For now, esp8266 is the only platform that supports hardware timers and
both hard and soft callbacks.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
docs: Document the cross-port Timer hard= option.

Update the main machine.Timer specification, and any references to
hard/soft interrupts in port-specific documentation. There is a separate
copy of the machine.Timer documentation for the pyboard, so update that
too to keep everything consistent.

Signed-off-by: Chris Webb <chris@arachsys.com>

arachsys committed:
shared/runtime/mpirq: Check separate hard IRQ stack correctly.

On the zephyr port, hard IRQ handlers run with a separate stack on a
different thread, so each call to mp_irq_dispatch() and mp_irq_handler()
has to be wrapped with adjustments to the stack-limit checker.

Move these adjustments into the shared mp_irq_dispatch(), introducing
MICROPY_STACK_SIZE_HARD_IRQ which a port can define to non-zero if it
uses a separate stack for hard IRQ handlers. We only need wrap the hard
dispatch case. This should reduce binary size on zephyr without affecting
other ports.

Signed-off-by: Chris Webb <chris@arachsys.com>

dpgeorge commented on tests/extmod/machine_timer.py:
```diff
@@ -0,0 +1,48 @@
+
+for mode in modes:
+    for kind in kinds:
+        for freq in 500, 250:
```
It might be a good idea to (eventually) add a second timer test to check for more fine-grained accuracy of the period/frequency.  Eg using just hard=True, and store into a pre-allocated list the time of each callback.  Then at the end make sure those times were all within a certain percentage of the set period.

dpgeorge reviewed (approved):
This is looking really good now.  I tested on esp32, esp8266, stm32 (PYBV10), rp2 and zephyr and the timer tests work as expected.

Overall this is a neat clean up and consolidation of code, and makes the ports more consistent.  And thanks for making well defined commits with good commit comments, and including docs and tests updates.  Very good!
